File: backend/app/schemas/traffic.py
"""Pydantic v2 schemas for the read-only `/v1/traffic` domain (live NSW
traffic cameras + hazards) — see `app.models.traffic` for the underlying
tables and `app.services.live_traffic` for where the data actually comes
from."""
from __future__ import annotations

from datetime import datetime
from typing import Generic, TypeVar

from pydantic import BaseModel, ConfigDict

# --- Pagination (local to this domain, same shape as app.schemas.geofence.Page) ---

T = TypeVar("T")


class Page(BaseModel, Generic[T]):
    items: list[T]
    total: int
    skip: int
    limit: int


class TrafficCameraRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    name: str
    latitude: float
    longitude: float
    direction: str | None
    image_url: str
    region: str | None


class TrafficHazardRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    category: str
    latitude: float
    longitude: float
    headline: str | None
    closure_type: str | None
    direction: str | None
    speed_limit: int | None
    expected_delay_minutes: int | None
    ended: bool
    source_last_published: datetime
    raw_json: dict

File: android/app/src/main/java/au/com/threesixty/cabdispatch/domain/Traffic.kt
package au.com.threesixty.cabdispatch.domain

/**
 * On-device copies of the two live NSW traffic tables (`GET /v1/traffic/cameras`,
 * `GET /v1/traffic/hazards`) — see [au.com.threesixty.cabdispatch.sync.TrafficCache], the Room
 * cache that populates these, and [au.com.threesixty.cabdispatch.ui.screens.hired.MeterBackdropMap],
 * the one place they are drawn (small, custom, non-interactive markers — informational only,
 * never a fare input). Same role [AirportZone]/[au.com.threesixty.cabdispatch.domain.fare.TollGantryRef]
 * play for their own cached reference data: a plain domain shape with no Room/Retrofit dependency,
 * so a caller never has to think in entity/DTO terms.
 */
data class TrafficCamera(
    val id: String,
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val direction: String? = null,
    val imageUrl: String? = null,
    val region: String? = null,
)

/** `app.models.traffic.TRAFFIC_HAZARD_CATEGORIES` — the real, fixed set of `category` values the
 * backend guarantees. Not an enum on this DTO/domain type (the backend is the source of truth for
 * what's valid, and a stale build must not crash on a category this list hasn't caught up to
 * yet) — kept here purely so [au.com.threesixty.cabdispatch.ui.screens.hired.MeterBackdropMap]'s
 * icon-colour choice has one documented place to point at rather than a magic string list. */
object TrafficHazardCategories {
    const val INCIDENT = "incident"
    const val ROADWORK = "roadwork"
    const val FLOOD = "flood"
    const val FIRE = "fire"
    const val ALPINE = "alpine"
    const val MAJOR_EVENT = "majorevent"
}

/** One live NSW hazard — incident, roadwork, flood, fire, alpine closure or major-event closure. */
data class TrafficHazard(
    val id: String,
    val category: String,
    val latitude: Double,
    val longitude: Double,
    val headline: String? = null,
    val closureType: String? = null,
    val direction: String? = null,
    val speedLimit: Int? = null,
    val expectedDelayMinutes: Int? = null,
)
